### DISCUSSION TITLE
C#: foreach breaking when the IEnumerator.MoveNext method throws exceptions

### DISCUSSION THREAD

**Discussion by YashojaLakmith** [dotnet/csharplang#7819] [Language Ideas] [answered]
### Description

When you iterate through a collection using the `foreach` and if there is a member in the collection which cannot be enumerated, the loop throws an exception and there's no way to skip that element or continue the loop even after catching the exception. The exception is thrown by the `IEnumerable.MoveNext()` method.

### Reproduction Steps

It is difficult to provide an example because, usually exception causing member is unexpected in the collection. One of the possible sources could be a enumerable returned from the `System.DirectoryInfo.EnumerateFiles()` method.

```
IEnumerable<FileInfo> files = new DirectoryInfo("D:").EnumerateFiles();

foreach (var file in files)  // Possible to throw an exception if there is an issue with a file path.
{
    // Process members
}
```

### Expected behavior

It is expected to be able to catch any exception that being thrown and handle them necessarily instead of breaking the iteration.

### Actual behavior

Exceptions thrown from `foreach` (which internally thrown by `IEnumerator.MoveNext()` method) cannot be handled without breaking the iteration of the loop. The exception being thrown is usually specific to the member in the enumerable. In the provided example, it could be file related exception such as `UnauthorizedAccessException`.

### Known Workarounds

The workaround is manually using IEnumerator.Movenext() method and IEnumerator<T>.Current property.
```
IEnumerable<FileInfo> files = new DirectoryInfo("D:").EnumerateFiles();
IEnumerator<FileInfo> e = files.GetEnumerator();

while (true)
{
    try
    {
        if (e.MoveNext())
        {
            // Process the member
        }
        else
        {
            break;
        }
    }
    catch(AggregateException ex)
    {
        // Handle exceptions
        continue;
    }
}
```

### Configuration

This behavior was seen in .NET 6, 7 and 8
OS: Windows 10 22H2, Windows 11 22H2 and Windows 11 23H2

**Comment by ghost** [comment]
Tagging subscribers to this area: @dotnet/area-system-collections
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.
<details>
<summary>Issue Details</summary>
<hr />

### Description

When you iterate through a collection using the `foreach` and if there is a member in the collection which cannot be enumerated, the loop throws an exception and there's no way to skip that element or continue the loop even after catching the exception. The exception is thrown by the `IEnumerable.MoveNext()` method.

### Reproduction Steps

It is difficult to provide an example because, usually exception causing member is unexpected in the collection. One of the possible sources could be a enumerable returned from the `System.DirectoryInfo.EnumerateFiles()` method.

```
IEnumerable<FileInfo> files = new DirectoryInfo("D:").EnumerateFiles();

foreach (var file in files)  // Possible to throw an exception if there is an issue with a file path.
{
    // Process members
}
```

### Expected behavior

It is expected to be able to catch any exception that being thrown and handle them necessarily instead of breaking the iteration.

### Actual behavior

Exceptions thrown from `foreach` (which internally thrown by `IEnumerator.MoveNext()` method) cannot be handled without breaking the iteration of the loop. The exception being thrown is usually specific to the member in the enumerable. In the provided example, it could be file related exception such as `UnauthorizedAccessException`.


### Regression?

_No response_

### Known Workarounds

The workaround is manually using IEnumerator.Movenext() method and IEnumerator<T>.Current property.
```
IEnumerable<FileInfo> files = new DirectoryInfo("D:").EnumerateFiles();
IEnumerator<FileInfo> e = files.GetEnumerator();

while (true)
{
    try
    {
        if (e.MoveNext())
        {
            // Process the member
        }
        else
        {
            break;
        }
    }
    catch(AggregateException ex)
    {
        // Handle exceptions
        continue;
    }
}
```

### Configuration

This behavior was seen in .NET 6, 7 and 8
OS: Windows 10 22H2, Windows 11 22H2 and Windows 11 23H2

### Other information

_No response_

<table>
  <tr>
    <th align="left">Author:</th>
    <td>YashojaLakmith</td>
  </tr>
  <tr>
    <th align="left">Assignees:</th>
    <td>-</td>
  </tr>
  <tr>
    <th align="left">Labels:</th>
    <td>

`area-System.Collections`, `untriaged`

</td>
  </tr>
  <tr>
    <th align="left">Milestone:</th>
    <td>-</td>
  </tr>
</table>
</details>

**Comment by huoyaoyuan** [comment]
I'm not sure whether the enumerator allows re-entrance when `MoveNext()` throws. The documentation for IEnuemerator doesn't mention that.

**Comment by HaloFour** [comment, marked as answer]
Given that this is an uncommon thing to want to do, having to rely on manual enumeration rather than `foreach` feels like a sufficient way to address this.

I would also expect that most enumerables aren't able to recover from an exception being thrown like this.

**Reply by NN---** [reply]
The similar behavior for await foreach where a single task can and should throw OperationCancelledException upon cancellation but this stops the whole iteration.
